Downgrade Tril's minimum compiler version

rbanerjee opened the pull request:
There was some desire to use Tril in contexts where the compiler may not have full C++11 support.
This adds that support to tril and it's accompanying examples, while still retaining some of the C++11 only code (by guarding them with appropriate macros).

rbanerjee commented on fvtest/tril/tril/ast.hpp:
```diff
@@ -28,6 +28,25 @@
 #include <cstring>
 
+#if __cplusplus <= 199711L
+const class nullptr_t {
```
This is kind of silly but nullptr was only added as recent as C++11. And just using NULL seems like a very crude solution (maybe I'm missing something here).
Open to suggestions as to whether there's a more elegant workaround.

mgaudet replied:
I'll admit to my gut not having a good feeling about what the 'right' answer here is, but, I would suggest this might be best put in another header; I don't think it makes sense here.

charliegracie replied:
I would really prefer if `NULL` was used.  Whenever possible we should try to be consistent throughout the entire codebase.

rbanerjee commented on fvtest/tril/examples/mandelbrot/main.cpp:
```diff
@@ -45,8 +49,13 @@ int main(int argc, char const * const * const argv) {
+    const auto size = 80;                   // number of rows/columns in the output table
+    const auto iterations = 1000;           // number of iterations to be performed
+#else
     constexpr auto size = 80;                   // number of rows/columns in the output table
```
Not sure why we are using constexpr here (instead of just const).

mgaudet replied:
I think I'd just keep the const auto versions.

mgaudet commented on fvtest/tril/tril/ast.hpp:
```diff
@@ -76,13 +101,17 @@ struct ASTValue {
      * is_String_Compatible<const char*>::value;   // true
      */
+#if __cplusplus <= 199711L
+     // Never used for the version without C++11 support.
```
I'd rather flip the sense of this check, and avoid the 

```
#if FOO 
//nothing
#else
stuff
#endif
```
pattern. It can be replaced with 

```
#if !FOO
stuff
#endif
```

mgaudet commented on fvtest/tril/tril/ast.hpp:
```diff
@@ -155,6 +213,21 @@ struct ASTValue {
      * b.isCompatibleWith<double>(); // returns true
      * b.isCompatibleWith<int>();    // returns false
      */
+#if __cplusplus <= 199711L
```
So, there's two distinctions here: 

1. The use of `enable_if` rather than `std::enable_if` 
2. The use of `std::is_integral` directly, rather than via the alias `is_Integer_Compatible` 

@Leonardo2718 Why the indirection of is_Integer_Compatible? 

@rbanerjee If we were to to standardize on `std::is_integral`, this could collapse nicely with a `using std::enable_if<T>` (syntax may be wrong).

Leonardo2718 replied:
The indirection is there mostly for two reasons:

1. In future, I was planning on making some changes to the type system in Tril, so having the indirection meant only having to change the code in one place.
2. Conceptually, I wanted there to be a distinction between the Tril type system and the C++ type system. The indirection was part of the interface for "bridging" the two, the goal being to hide as much of the details of how types are handled in the background behind a clear and precise interface.

Although I would prefer to keep the extra indirection for the sake of mental clarity, I'm okay with removing it as part of this change. In this case, I see no value in keeping both implementations, since it would defeat the goal of reducing code duplication.

mgaudet commented on fvtest/tril/tril/ilgen.hpp:
```diff
@@ -50,7 +50,11 @@ class TRLangBuilder : public TR::IlInjector {
         TRLangBuilder(const ASTNode* trees, TR::TypeDictionary* d)
               : TR::IlInjector(d), _trees(trees) {}
 
+#if __cplusplus <= 199711L // Check compiler support for C++11.
```
For changes like this, since override is not mandatory, I think I'd prefer to skip the `#ifdef` and instead simply annotate the function with a comment;

```
bool injectIL() /* override */ ;
```

mgaudet replied:
Same applies elsewhere

mgaudet commented on fvtest/tril/examples/mandelbrot/main.cpp:
```diff
@@ -25,7 +25,11 @@
 #include <assert.h>
 #include <stdio.h>
 
+#if __cplusplus <= 199711L
```
I think I'd just keep the typedef, rather than #ifdef.

mgaudet commented on fvtest/tril/tril/ast.hpp:
```diff
@@ -59,9 +78,15 @@ struct ASTValue {
 
     // aliases for the underlying C++ types used to store the different AST types
+#if __cplusplus <= 199711L // Check for C++11 compiler support.
+    typedef uint64_t Integer_t;
```
I think I'd just keep the typedefs, rather than ifdefs. It doesn't bring extra value preserving the the `using`

Leonardo2718 commented on fvtest/tril/tril/ilgen.cpp:
```diff
@@ -45,7 +45,13 @@ class OpCodeTable : public TR::ILOpCode {
          if (opcode == _opcodeNameMap.cend()) {
+#if __cplusplus <= 199711L
+            for (int i = TR::FirstOMROp; i< TR::NumIlOps; i++) {
+               TR::ILOpCodes p_opCode = (TR::ILOpCodes)i;
```
I would prefer `const auto p_opCode = static_cast<TR::ILOpCodes>(i);`

Leonardo2718 commented on fvtest/tril/tril/ilgen.cpp:
```diff
@@ -45,7 +45,13 @@ class OpCodeTable : public TR::ILOpCode {
+            for (int i = TR::FirstOMROp; i< TR::NumIlOps; i++) {
+               TR::ILOpCodes p_opCode = (TR::ILOpCodes)i;
+               OMR::OpCodeProperties &p = TR::ILOpCode::_opCodeProperties[p_opCode];
+#else
```
I would prefer `const auto& p = TR::ILOpCode::_opCodeProperties[p_opCode];`

Leonardo2718 commented on fvtest/tril/tril/ilgen.cpp:
```diff
@@ -45,7 +45,13 @@ class OpCodeTable : public TR::ILOpCode {
+               TR::ILOpCodes p_opCode = (TR::ILOpCodes)i;
+               OMR::OpCodeProperties &p = TR::ILOpCode::_opCodeProperties[p_opCode];
+#else
             for (const auto& p : _opCodeProperties) {
```
Not sure there's any value in having both implementations.

rbanerjee commented:
@Leonardo2718 , @mgaudet 

> Although a theoretically sound approach to checking for C++11 support, in practice, using #if __cplusplus <= 199711L is not usually considered safe/portable because some compilers have __cplusplus set to 199711L even if C++11 support is incomplete."

About using the __cplusplus macro. There's actually no "correct" way of doing this it seems.

"Implementations rarely leap from one formal revision of the standard directly to the next; the implementation process generally proceeds by smaller steps. As a result, testing for a specific revision of the standard (e.g. by examining the value of the __cplusplus macro) often gives the wrong answer. Implementers generally don’t want to appear to be claiming full conformance to a standard revision until all of its features are implemented. That leaves programmers with no portable way to determine which features are actually available to them."

^ this is from https://isocpp.org/std/standing-documents/sd-6-sg10-feature-test-recommendations#recs.hasinc

Which is partly why I left most of Leo's old code behind (the ones which strictly require C++11 support) while guarding them with a macro. Because at IMO,  in most cases they are doing the right thing (given the C++11 tools are actually available). 

My idea was to remove the "work arounds" once we actually have a C++11 supported build compiler. So I was kind of surprised as to why we would want to get rid of all the C++11 code altogether (cause in theory at least, they are "better"). This also kind of answers the questions about "Not sure if there's any value in having both implementations of X?". Unfortunately though, what I was unaware of was how often we updated our build compilers to begin with. 

So given it might take a while for was to switch to a build compiler that actually has C++11 support, I think it does make sense as to just get rid of things like std::enable_if, lambdas and all the usings.

 At least that's my understanding of the situation. Please do correct me if I'm wrong.

rbanerjee commented on fvtest/tril/tril/ast.hpp:
```diff
@@ -108,6 +137,34 @@ struct ASTValue {
+#if __cplusplus <= 199711L
+    // Create create an equivalent implementation of std::enable_if.
+    template <bool, typename T = void>
+    struct enable_if {};
```
It doesn't (at least I couldn't get it to work on GCC 4.4.6 [also, notice it says *at least as far back as GCC 4.4.6] by default).

But more importantly one of the original reasons why are doing this to begin with, to quote @fjeremic was 

>In particular attempting to build Tril on Linux on z with g++ 4.4.6 will not work due to missing C++11 support.

So instead of getting it to compile with some particular version of GCC, this patch is supposed to remove/work around the C++11 only features. And as far as I know, std::enable_if was "officially" only added since C++11.

mgaudet replied:
Did you end up compiling Tril (You should check the command lines) w/o any C++11 support? Or just whatever comes out of GCC 4.4.6 (i.e., `-std=c++0x`)?

rbanerjee replied:
Can confirm, doesn't work.

mgaudet commented:
Personally, I'd like to keep the supported subset that the rest of the compiler is already using -- It'd be silly to get rid of `auto` for example, given that the rest of the compiler already uses `auto`.

charliegracie commented:
It may be a really long time before we have a C++11 compiler available on all of our platforms. To provide a consistent environment for all platforms weave to use the LCD. 

I agree with @mgaudet on keeping the features which the rest of the compiler already uses. I wish his example was not 'auto' though as it I should one of the worst features c++ added in my opinion. It makes debugging and code inspection a nightmare but that is a bridge that was already crossed so we are stuck with it.

mgaudet commented:
Heh. It's easily my favourite!

rbanerjee commented:
ptal

mgaudet commented on fvtest/tril/tril/jitbuilder_compiler.cpp:
```diff
@@ -40,6 +40,12 @@ int32_t Tril::JitBuilderCompiler::compile() {
    return compileWithVerifier(NULL);
 }
 
+#if __cplusplus <= 199711L // Check for C++11 compiler support.
```
Given the way the rest of this change looks, I'd just pull this, and use the lambda version universally.

rbanerjee replied:
Sorry, but did you mean the 'version *without the lambda universally'? (Lambdas were again only added since C++11).

mgaudet replied:
I'm cool with this.

Leonardo2718 commented on fvtest/tril/tril/enable_if.hpp:
```diff
@@ -0,0 +1,40 @@
+    // Create an equivalent implementation of std::enable_if.
+    template <bool, typename T = void>
+
+    struct enable_if {};
```
Can you put this in the `Tril` namespace?

rbanerjee replied:
Done.

rbanerjee commented:
ptal

lmaisons commented on fvtest/tril/tril/enable_if.hpp:
```diff
@@ -0,0 +1,46 @@
+#ifndef ENABLE_IF_HPP
+#define ENABLE_IF_HPP
+
+#if __cplusplus > 199711L
```
This version check seems somewhat aggressive to me, I would have thought we would want something more along the lines of
```
#if (__cplusplus < 201103L) || !defined(_MSVC_LANG)
```
with `_MSVC_LANG` assuming that MSVC++2015+ is acceptable...

Do we know for sure that all partial updates to the C++ language standard on the compilers we care about either
1. Don't update the _cplusplus version or,
2. Are compatible with our requirements?

rbanerjee replied:
@lmaisons 

This was brought up in an earlier comment, but I don't think there's any way of coming up with a combination of macros that is guaranteed to be infallible when trying to detect whether the compiler does or doesn't have C++11 support [checking `__cplusplus` by itself *theoretically should be enough] . 

>"Implementations rarely leap from one formal revision of the standard directly to the next; the implementation process generally proceeds by smaller steps. As a result, testing for a specific revision of the standard (e.g. by examining the value of the __cplusplus macro) often gives the wrong answer. Implementers generally don’t want to appear to be claiming full conformance to a standard revision until all of its features are implemented. That leaves programmers with no portable way to determine which features are actually available to them."

 [ ref- https://isocpp.org/std/standing-documents/sd-6-sg10-feature-test-recommendations#recs.hasinc . I guess C++17 has `__has_cpp_attribute` which kind of does the trick, but of course that's no use to us.  ]

Right now how it's set up, it at least works for our build compiler and the newest GCC one. Not sure how much marginal benefit we might potentially be able to extract from complicating the `#if` statements and whether it's even "worth" doing that. I might be missing something here.

rbanerjee commented:
@mgaudet  @0xdaryl  

Think this is good to merge.

fjeremic commented:
Any news on this item? We have some pending work that is blocked by this.

dchopra001 commented:
I see the following on gcc 4.4.6 on zLinux:

```[100%] Building CXX object fvtest/compilertriltest/CMakeFiles/comptest.dir/__/omrGtestGlue/argmain.cpp.o
In file included from /gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/OpCodeTest.hpp:25,
                 from /gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/ArithmeticTest.cpp:22:
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:221: error: expected constructor, destructor, or type conversion before ‘T’
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:222: error: expected constructor, destructor, or type conversion before ‘T’
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:223: error: expected constructor, destructor, or type conversion before ‘T’
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:224: error: expected constructor, destructor, or type conversion before ‘T’
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:225: error: expected constructor, destructor, or type conversion before ‘T’
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp: In function ‘std::vector<T, std::allocator<_CharT> > TRTest::const_values()’:
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:233: error: ‘zero_value’ was not declared in this scope
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:234: error: ‘one_value’ was not declared in this scope
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:235: error: ‘negative_one_value’ was not declared in this scope
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:236: error: ‘positive_value’ was not declared in this scope
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:237: error: ‘negative_value’ was not declared in this scope
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:233: error: expected primary-expression before ‘{’ token
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:233: error: expected ‘;’ before ‘{’ token
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:233: error: expected primary-expression before ‘>’ token
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:233: error: expected primary-expression before ‘)’ token
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:234: error: expected primary-expression before ‘>’ token
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:234: error: expected primary-expression before ‘)’ token
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:235: error: expected primary-expression before ‘>’ token
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:235: error: expected primary-expression before ‘)’ token
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:236: error: expected primary-expression before ‘>’ token
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:236: error: expected primary-expression before ‘)’ token
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:237: error: expected primary-expression before ‘>’ token
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:237: error: expected primary-expression before ‘)’ token
/gsa/tlbgsa/home/d/c/dchopra/development/omr/fvtest/compilertriltest/JitTest.hpp:242: error: expected ‘;’ before ‘}’ token
...
```

rbanerjee commented:
Looking into it now.

rbanerjee commented:
@fjeremic  So with this PR, it seems like `Tril` and the examples work fine on `zLinux` but some of the Tril Compiler Tests still need to be ported. As the original Issue #1697 suggested, should we do this in two parts?

As for this one, it takes care of getting Tril, the parser and the Ilgenerator up and running with older compilers. I think it's best we merge this and I can open up a separate PR for porting the tests, as there is a certain amount of discussion to be had there. Does that sound good?

rbanerjee commented:
Just to confirm, the blocked PR is #1557, right?

fjeremic commented:
> @fjeremic Just to confirm, the blocked PR is #1557, right?

Correct.

rbanerjee commented:
> So with this PR, it seems like Tril and the examples work fine on zLinux but some of the Tril Compiler Tests still need to be ported. As the original Issue #1697 suggested, should we do this in two parts?

>As for this one, it takes care of getting Tril, the parser and the Ilgenerator up and running with older compilers. I think it's best we merge this and I can open up a separate PR for porting the tests, as there is a certain amount of discussion to be had there. Does that sound good?

@fjeremic Do you have any thoughts on this part?

@0xdaryl  ptal. If there aren't any objections then I would l like to have this merged please.

fjeremic commented:
> @fjeremic Do you have any thoughts on this part?

Sounds reasonable to me.

0xdaryl commented:
omr-genie build all

0xdaryl commented:
genie-omr build all

charliegracie commented:
I would prefer to see all C++11 code completely removed as I see no win in maintaining both copies. I think this is agreeing with @0xdaryl.  Other components have sadly let C++0x extensions creep in so that is ok but until C++11 is fully supported on all platforms and is a specification level we can target I think removing it is the best course of action.

rbanerjee commented:
@0xdaryl  Addressed the comments. ptal.

>I would prefer to see all C++11 code completely removed as I see no win in maintaining both copies. I think this is agreeing with @0xdaryl. Other components have sadly let C++0x extensions creep in so that is ok but until C++11 is fully supported on all platforms and is a specification level we can target I think removing it is the best course of action.

@charliegracie The checking of `__cplusplus <= 199711L ` macros have now completely been removed in this patchset. Actually there were only two of them left form the last iteration. 

One for using `std::enable_if` instead of our own implementation. IIUC  using the one  in `std` doesn't really provide any added benefits. So I don't have any issues removing that one.

The other one had to do with using `std:: bind1st` and `std::ptr_fun` to "emulate" lambdas. The issue with that one being, from my understanding `std::bind1st` has been deprecated since C++11, and C++17 completely removes it. So once we upgrade our compilers, having only the implementation with `std::bind1st` should break our builds. However, don't think we are going to be moving to C++17 complaint compilers anytime in the near future. So either I can open up an issue to document this behaviour, or we can just use the macro for that one specific case?

mgaudet commented on fvtest/tril/tril/jitbuilder_compiler.cpp:
```diff
@@ -51,7 +55,7 @@ int32_t Tril::JitBuilderCompiler::compileWithVerifier(TR::IlVerifier* verifier)
     auto argIlTypes = std::vector<TR::IlType*>{argTypes.size()};
     std::transform(argTypes.cbegin(), argTypes.cend(), argIlTypes.begin(),
-                   [&types](TR::DataTypes d){ return types.PrimitiveType(d); } );
+                   std::bind1st(std::ptr_fun(&getPrimitiveType),types));
```
(We discussed this in person) 

Given the simplicity of `std::transform`, IMO, rather than chasing the 'functional purity', it'd be better to just do the loop, which removes the need to have a function defined, and eliminates both `std::bind1st` and `std::ptr_fun`.

rbanerjee commented:
@0xdaryl  ptal. Addressed the comments and testing is done. (Removing the WIP tag.)

0xdaryl commented:
omr-genie build all

0xdaryl commented:
genie-omr build all

0xdaryl commented:
I believe all concerns have been addressed.  There are no compiler level macros remaining in this PR.
